File: numbers/src/main/java/org/dicio/numbers/unit/Number.java
package org.dicio.numbers.unit;

import java.util.Objects;

/**
 * TODO add documentation
 */
public class Number {

    private final boolean isDecimal;
    private final long integerValue;
    private final double decimalValue;
    private final boolean isOrdinal;


    public Number(final long integerValue) {
        this(integerValue, false);
    }

    public Number(final double decimalValue) {
        this(decimalValue, false);
    }

    public Number(final long integerValue, final boolean isOrdinal) {
        this(false, integerValue, Double.NaN, isOrdinal);
    }

    public Number(final double decimalValue, final boolean isOrdinal) {
        this(true, 0, decimalValue, isOrdinal);
    }

    private Number(final boolean isDecimal,
                   final long integerValue,
                   final double decimalValue,
                   final boolean isOrdinal) {
        this.isDecimal = isDecimal;
        this.integerValue = integerValue;
        this.decimalValue = decimalValue;
        this.isOrdinal = isOrdinal;
    }


    public static Number fromObject(final Object object) {
        if (object instanceof Short || object instanceof Integer || object instanceof Long) {
            return new Number(((java.lang.Number) object).longValue());
        } else if (object instanceof Float || object instanceof Double) {
            return new Number(((java.lang.Number) object).doubleValue());
        } else {
            throw new IllegalArgumentException(
                    "object is neither an integer nor a decimal number: " + object);
        }
    }


    public boolean isDecimal() {
        return isDecimal;
    }

    public boolean isInteger() {
        return !isDecimal;
    }

    public long integerValue() {
        return integerValue;
    }

    public double decimalValue() {
        return decimalValue;
    }

    public boolean isOrdinal() {
        return isOrdinal;
    }

    public Number withOrdinal(final boolean isOrdinal) {
        return new Number(this.isDecimal, this.integerValue, this.decimalValue, isOrdinal);
    }


    public Number multiply(final long integer) {
        if (isDecimal) {
            return new Number(decimalValue * integer, isOrdinal);
        } else {
            try {
                return new Number(Math.multiplyExact(integerValue, integer), isOrdinal);
            } catch (final ArithmeticException e) {
                return new Number((double) integerValue * integer, isOrdinal);
            }
        }
    }

    public Number multiply(final double decimal) {
        return new Number((isDecimal ? decimalValue : integerValue) * decimal, isOrdinal);
    }

    public Number multiply(final Number number) {
        return (number.isDecimal ? multiply(number.decimalValue) : multiply(number.integerValue));
    }

    public Number plus(final long integer) {
        if (isDecimal) {
            return new Number(decimalValue + integer, isOrdinal);
        } else {
            try {
                return new Number(Math.addExact(integerValue, integer), isOrdinal);
            } catch (final ArithmeticException e) {
                return new Number((double) integerValue + integer, isOrdinal);
            }
        }
    }

    public Number plus(final double decimal) {
        return new Number((isDecimal ? decimalValue : integerValue) + decimal, isOrdinal);
    }

    public Number plus(final Number number) {
        return number.isDecimal ? plus(number.decimalValue) : plus(number.integerValue);
    }

    public Number divide(final long integer) {
        if (isDecimal) {
            return new Number(decimalValue / integer, isOrdinal);
        } else if (integerValue % integer == 0) {
            return new Number(integerValue / integer, isOrdinal);
        } else {
            return new Number(((double) integerValue) / integer, isOrdinal);
        }
    }

    public Number divide(final double decimal) {
        return new Number((isDecimal ? decimalValue : integerValue) / decimal, isOrdinal);
    }

    public Number divide(final Number number) {
        return number.isDecimal ? divide(number.decimalValue) : divide(number.integerValue);
    }

    public boolean lessThan(final long integer) {
        return isDecimal ? (decimalValue < integer) : (integerValue < integer);
    }

    public boolean lessThan(final double decimal) {
        return isDecimal ? (decimalValue < decimal) : (integerValue < decimal);
    }

    public boolean moreThan(final long integer) {
        return isDecimal ? (decimalValue > integer) : (integerValue > integer);
    }


    public boolean equals(final long integer) {
        return !isDecimal && integerValue == integer;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        } else if (o == null || getClass() != o.getClass()) {
            return false;
        } else {
            final Number number = (Number) o;
            return isDecimal == number.isDecimal && isOrdinal == number.isOrdinal && (isDecimal
                    ? decimalValue == number.decimalValue : integerValue == number.integerValue);
        }
    }

    @Override
    public int hashCode() {
        return isDecimal ? Objects.hash(true, decimalValue) : Objects.hash(false, integerValue);
    }

    @Override
    public String toString() {
        return (isDecimal ? String.valueOf(decimalValue) : String.valueOf(integerValue))
                + (isOrdinal ? "th" : "");
    }
}

File: numbers/src/main/java/org/dicio/numbers/unit/Number.kt
package org.dicio.numbers.unit

import java.util.Objects

/**
 * TODO add documentation
 */
class Number private constructor(
    @JvmField val isDecimal: Boolean,
    private val integerValue: Long,
    private val decimalValue: Double,
    val isOrdinal: Boolean
) {
    @JvmOverloads
    constructor(integerValue: Long, isOrdinal: Boolean = false) : this(
        false,
        integerValue,
        Double.NaN,
        isOrdinal
    )

    @JvmOverloads
    constructor(decimalValue: Double, isOrdinal: Boolean = false) : this(
        true,
        0,
        decimalValue,
        isOrdinal
    )


    val isInteger: Boolean
        get() = !isDecimal

    fun integerValue(): Long {
        return integerValue
    }

    fun decimalValue(): Double {
        return decimalValue
    }

    fun withOrdinal(isOrdinal: Boolean): Number {
        return Number(this.isDecimal, this.integerValue, this.decimalValue, isOrdinal)
    }


    fun multiply(integer: Long): Number {
        return if (isDecimal) {
            Number(decimalValue * integer, isOrdinal)
        } else {
            try {
                Number(Math.multiplyExact(integerValue, integer), isOrdinal)
            } catch (e: ArithmeticException) {
                Number(integerValue.toDouble() * integer, isOrdinal)
            }
        }
    }

    fun multiply(decimal: Double): Number {
        return Number(
            (if (isDecimal) decimalValue else integerValue.toDouble()) * decimal,
            isOrdinal
        )
    }

    fun multiply(number: Number?): Number {
        return (if (number!!.isDecimal) multiply(number.decimalValue) else multiply(
            number.integerValue
        ))
    }

    fun plus(integer: Long): Number {
        return if (isDecimal) {
            Number(decimalValue + integer, isOrdinal)
        } else {
            try {
                Number(Math.addExact(integerValue, integer), isOrdinal)
            } catch (e: ArithmeticException) {
                Number(integerValue.toDouble() + integer, isOrdinal)
            }
        }
    }

    fun plus(decimal: Double): Number {
        return Number(
            (if (isDecimal) decimalValue else integerValue.toDouble()) + decimal,
            isOrdinal
        )
    }

    fun plus(number: Number?): Number {
        return if (number!!.isDecimal) plus(number.decimalValue) else plus(number.integerValue)
    }

    fun divide(integer: Long): Number {
        return if (isDecimal) {
            Number(decimalValue / integer, isOrdinal)
        } else if (integerValue % integer == 0L) {
            Number(integerValue / integer, isOrdinal)
        } else {
            Number((integerValue.toDouble()) / integer, isOrdinal)
        }
    }

    fun divide(decimal: Double): Number {
        return Number(
            (if (isDecimal) decimalValue else integerValue.toDouble()) / decimal,
            isOrdinal
        )
    }

    fun divide(number: Number): Number {
        return if (number.isDecimal) divide(number.decimalValue) else divide(number.integerValue)
    }

    fun lessThan(integer: Long): Boolean {
        return if (isDecimal) (decimalValue < integer) else (integerValue < integer)
    }

    fun lessThan(decimal: Double): Boolean {
        return if (isDecimal) (decimalValue < decimal) else (integerValue < decimal)
    }

    fun moreThan(integer: Long): Boolean {
        return if (isDecimal) (decimalValue > integer) else (integerValue > integer)
    }


    fun equals(integer: Long): Boolean {
        return !isDecimal && integerValue == integer
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        } else if (other == null || javaClass != other.javaClass) {
            return false
        } else {
            val number = other as Number
            return isDecimal == number.isDecimal && isOrdinal == number.isOrdinal && (if (isDecimal
            ) decimalValue == number.decimalValue else integerValue == number.integerValue)
        }
    }

    override fun hashCode(): Int {
        return if (isDecimal) Objects.hash(true, decimalValue) else Objects.hash(
            false,
            integerValue
        )
    }

    override fun toString(): String {
        return ((if (isDecimal) decimalValue.toString() else integerValue.toString())
                + (if (isOrdinal) "th" else ""))
    }

    companion object {
        fun fromObject(o: Any): Number {
            return when (o) {
                is Short, is Int, is Long -> Number((o as kotlin.Number).toLong())
                is Float, is Double -> Number((o as kotlin.Number).toDouble())
                else -> throw IllegalArgumentException(
                        "object is neither an integer nor a decimal number: $o")
            }
        }
    }
}
